pkg/agent/multicast/mcast_discovery.go

```diff
@@ -97,14 +107,68 @@ func (s *IGMPSnooper) queryIGMP(group net.IP, versions []uint8) error {
 		if err != nil {
 			return err
 		}
-		if err := s.ofClient.SendIGMPQueryPacketOut(igmpQueryDstMac, mcastAllHosts, openflow13.P_NORMAL, igmp); err != nil {
+		outPort := uint32(0)
+		if err := s.ofClient.SendIGMPQueryPacketOut(igmpQueryDstMac, types.McastAllHosts, outPort, igmp); err != nil {
 			return err
 		}
-		klog.V(2).InfoS("Sent packetOut for IGMP query", "group", group.String(), "version", version)
+		klog.V(2).InfoS("Sent packetOut for IGMP query", "group", group.String(), "version", version, "outPort", outPort)
 	}
 	return nil
 }
 
+func (s *IGMPSnooper) validate(event *mcastGroupEvent) (bool, error) {
+	if s.validator == nil {
+		// Do not need to validate if the packet should be dropped since Validator is null, thus here returns directly
+		return true, nil
+	}
+	if event.iface.Type == interfacestore.ContainerInterface {
+		item, err := s.validator.Validate(event.iface.PodName, event.iface.PodNamespace, event.group)
+		klog.V(2).InfoS("call function s.Validator.Validate", "podName", event.iface.PodName,
+			"podNamespace", event.iface.PodNamespace, "mgroup", event.group.String())
+		if err != nil {
+			// Here just checks if packet should be dropped or not, Validate will return error when it fails to call ByIndex.
+			// We just imagine there is no rule for the pod(event.iface.PodName, event.iface.PodNamespace) to groupAddress event.group.
+			klog.ErrorS(err, "s.Validator.Validate returned")
+			return true, err
+		}
+		if item.NPType != nil {
+			klog.V(4).InfoS("s.Validator.Validate returned value", "RuleAction", *item.RuleAction,
+				"UUID", item.Uuid, "Name", item.Name)
+			s.addToIGMPReportNPStatsMap(item, event.packetLen)
+		}
+		if item.RuleAction != nil && *item.RuleAction == v1alpha1.RuleActionDrop {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func (s *IGMPSnooper) addToIGMPReportNPStatsMap(item types.McastNPValidationItem, packetLen uint64) {
+	uuid := item.Uuid
+	name := item.Name
+	ruleType := *item.NPType
+	s.igmpReportNPStatsMutex.Lock()
+	defer s.igmpReportNPStatsMutex.Unlock()
+	if s.igmpReportNPStats[ruleType][uuid] != nil {
+		if s.igmpReportNPStats[ruleType][uuid][name] != nil {
+			stats := s.igmpReportNPStats[ruleType][uuid][name]
+			stats.Bytes += packetLen
+			stats.Packets += 1
+		} else {
+			s.igmpReportNPStats[ruleType][uuid][name] = &types.RuleMetric{Bytes: packetLen, Packets: 1}
+		}
+	} else {
+		s.igmpReportNPStats[ruleType][uuid] = make(map[string]*types.RuleMetric)
+		s.igmpReportNPStats[ruleType][uuid][name] = &types.RuleMetric{Bytes: packetLen, Packets: 1}
+	}
+}
+
+func (s *IGMPSnooper) GetStatsMap() map[v1beta2.NetworkPolicyType]map[apitypes.UID]map[string]*types.RuleMetric {
+	s.igmpReportNPStatsMutex.RLock()
+	defer s.igmpReportNPStatsMutex.RUnlock()
```

Comment:
I do not think it helps. The lock is released after the func returns, and the caller will access the map without any lock.

Reply:
Please read https://github.com/antrea-io/antrea/pull/3449#discussion_r883522206